### PR TITLE
Added Support for More Options in SQL

### PR DESCRIPTION
**Pull request by AliabbasMerchant**

### CODE REVIEW

File: modules/crud/sql/read.go [hunk 1]
```diff
@@ -69,18 +69,30 @@ func (s *SQL) Read(ctx context.Context, project, col string, req *model.ReadRequ
...
+	case utils.Count:
+		query = query.Select(goqu.COUNT("*"))
+	case utils.Distinct:
+		distinct := req.Options.Distinct
+		if distinct == nil {
+			return nil, utils.ErrInvalidParams
+		}
+		query = query.SelectDistinct(*distinct)
+	case utils.One, utils.All:
+		query = query.Select(selArray...)
```
Comment (YourTechBud) [inline modules/crud/sql/read.go]: What if the sel array is empty,?

File: modules/crud/sql/read.go [hunk 2]
```diff
@@ -101,6 +113,27 @@ func (s *SQL) Read(ctx context.Context, project, col string, req *model.ReadRequ
...
+			return nil, err
+		}
+
+		switch s.GetDBType() {
+		case utils.MySQL, utils.Postgres:
+			mysqlTypeCheck(rowTypes, mapping)
+		}
+
+		for _, v := range mapping {
+			return v, nil
```
Comment (YourTechBud) [inline modules/crud/sql/read.go]: Shouldn't we return an object here? What does mongo return currently? We should make sql seem similar

File: modules/crud/sql/read.go [hunk 3]
```diff
@@ -128,11 +161,6 @@ func (s *SQL) Read(ctx context.Context, project, col string, req *model.ReadRequ
 				return nil, err
 			}
 
-			switch s.GetDBType() {
-			case utils.MySQL, utils.Postgres:
-				rowTypes, _ = rows.ColumnTypes()
```
Comment (YourTechBud) [inline modules/crud/sql/read.go]: Why did you remove this?

File: modules/crud/sql/update.go [hunk 4]
```diff
@@ -1,60 +1,226 @@
...
+		}
+	case utils.Upsert:
+		for k := range req.Update {
+			switch k {
+			case "$set", "$inc", "$mul", "$max", "$min", "$currentDate":
+				err := s.upsert(ctx, project, col, req, k)
+				if err != nil {
+					return err
+				}
+			default: // (case "$push", "$unset", "$rename")
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: Should we silently ignore this or throw an error? @Jayesh333 what do you think?

File: modules/crud/sql/update.go [hunk 5]
```diff
@@ -1,60 +1,226 @@
...
+	if req == nil {
+		return utils.ErrInvalidParams
 	}
-	return s.doExecContext(ctx, sqlString, args)
+	if req.Update == nil {
+		return utils.ErrInvalidParams
+	}
+	switch req.Operation {
+	case utils.All:
+		for k := range req.Update {
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: We shouldn't have a for loop here. `generateUpdateQuery` should internal have this for loop in my opinion.

File: modules/crud/sql/update.go [hunk 6]
```diff
@@ -1,60 +1,226 @@
...
 	sqlString = strings.Replace(sqlString, "\"", "", -1)
+	switch op {
+	case "$set":
+	case "$inc":
+		for k, v := range m {
+			val, err := numToString(v)
+			if err != nil {
+				return "", nil, err
+			}
+			sqlString = strings.Replace(sqlString, k+"="+val, k+"="+k+"+"+val, -1)
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: This is prone to injection attacks.
Why aren't we using a transaction? Such operations aren't truly atomic in all cases. We'll explicitly set the consistency level to make sure we have a record level lock so no other query can simultaneously perform a read or write.

File: modules/crud/sql/update.go [hunk 7]
```diff
@@ -1,60 +1,226 @@
...
+	case "$currentDate":
+		for k, v := range m {
+			val, ok := v.(string)
+			if !ok {
+				return "", nil, utils.ErrInvalidParams
+			}
+			if val == "timestamp" {
+				sqlString = strings.Replace(sqlString, k+"='"+val+"'", k+"=CURRENT_TIMESTAMP()", -1)
+			} else {
+				sqlString = strings.Replace(sqlString, k+"='"+val+"'", k+"=CURRENT_DATE()", -1)
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: Do we have `CURRENT_DATE()` and `CURRENT_TIMESTAMP()` exist for both, postgres and MySQL?

File: modules/crud/sql/update.go [hunk 8]
```diff
@@ -1,60 +1,226 @@
...
+	}
+	if rowsAffected == 0 {
+		if op == "$currentDate" {
+			return utils.ErrInvalidParams
+		}
+		sqlQuery, args, err := s.generateCreateQuery(ctx, project, col, &model.CreateRequest{Document: req.Update[op], Operation: utils.One})
+		if err != nil {
+			return err
+		}
+		err = doExecContext(ctx, sqlQuery, args, tx)
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: You are firing one create query for each update op. Seems to be wrong and it won't have goot performance

**Comment by YourTechBud** [comment]
@AliabbasMerchant did you make those changes?

**Comment by AliabbasMerchant** [comment]
@YourTechBud Done!

File: modules/crud/sql/update.go [hunk 9]
```diff
@@ -1,60 +1,236 @@
...
+			}
+			for op := range req.Update {
+				if op == "$currentDate" {
+					return utils.ErrInvalidParams
+				}
+				m, ok := req.Update[op].(map[string]interface{})
+				if !ok {
+					return utils.ErrInvalidParams
+				}
+				for k, v := range m { // k -> $max, $min, $inc, $mul
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: Need to correct the comment here

File: modules/crud/sql/update.go [hunk 10]
```diff
@@ -1,60 +1,236 @@
...
+		defer rows.Close()
+		
+		if !rows.Next() {
+			// not found. So, insert
+			doc := make(map[string]interface{})
+			for k, v := range req.Find {
+				doc[k] = v
+			}
+			for op := range req.Update {
+				if op == "$currentDate" {
```
Comment (YourTechBud) [inline modules/crud/sql/update.go]: We need to be able to support `$currentDate` in upserts too
